test/compile_errors.zig

```diff
@@ -3,6 +3,15 @@ const builtin = @import("builtin");
 const Target = @import("std").Target;
 
 pub fn addCases(cases: *tests.CompileErrorContext) void {
+    cases.addTest("access of undefined pointer to array",
```

Comment:
This test case isn't correct - `ram_u32` is a global variable, and it might have been reassigned before `entry` is executed, so it can't be a compile error detecting use of undefined here without some advanced analysis that Zig is not currently capable of. Also the "use of undefined value" as a compile error is deprecated (See #1947), only certain kinds of use of `undefined` is invalid, and it gives the message, "use of undefined value here causes undefined behavior", which should be emitted via a call to `ir_resolve_const_val` (or `ir_resolve_const`) with the `UndefAllowed` parameter set to `UndefBad`.

It's a bit tricky - when obtaining the pointer to a global variable, it is considered to be comptime known, however the memory that it points to is runtime known. So if `ir_analyze_instruction_elem_ptr` is seeing `ConstValSpecialUndef` then I think the bug has already occurred elsewhere. `ir_analyze_instruction_elem_ptr` should be seeing `ConstValSpecialRuntime`.

Reply:
Indeed, I got sidetracked by the complex logic and forgot that it was a `var`. The new patch allows the code to compile (no idea how to test that) and I've added a check to make sure the use of undefined is caught at compile-time.